File: templates/problems/07_trees/05_same_tree/python/solution.py
from helpers import TreeNode, Optional


def solve_is_same_tree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    """Return true if two binary trees are structurally identical.

    Time: O(n), Space: O(h).
    """
    if p is None and q is None:
        return True
    if p is None or q is None or p.val != q.val:
        return False
    return solve_is_same_tree(p.left, q.left) and solve_is_same_tree(p.right, q.right)

File: templates/problems/07_trees/05_same_tree/typescript/solution.ts
import { TreeNode } from "./helpers";

export function solveIsSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    if (p === null && q === null) {
        return true;
    }
    if (p === null || q === null || p.val !== q.val) {
        return false;
    }
    return solveIsSameTree(p.left, q.left) && solveIsSameTree(p.right, q.right);
}
